### PR TITLE
Upstreaming the Scalar command

### PR DESCRIPTION
**Pull request by dscho**
tl;dr: This series contributes the core part of the Scalar command to the Git project. This command provides a convenient way to clone/initialize very large repositories (think: monorepos).

Note: This patch series' focus is entirely on Scalar, on choosing sensible defaults and offering a delightful user experience around working with monorepos, and not about changing any existing paradigms for `contrib/` (even if catching up on the mail thread is likely to give interested readers that false impression).

Changes since v9:
- The patches to build Scalar and run its tests as part of Git's CI/PR, have been dropped because a recent unrelated patch series does not interact well with them.

Changes since v8:
- The rebase on top of v2.34.0, which changed the default merge strategy to ORT, should have changed the default for `merge.renames` to `true`. This is now the case.
- Accommodate preemptively for `ab/ci-updates` which invalidates assumptions made by this patch series that would still hold true with v2.34.0 but are no longer valid in `seen` and would trigger CI build breakages.

Changes since v7:
- Clarified in the commit message why we cannot easily encapsulate the Scalar part of the CMake configuration in `contrib/scalar/`.
- Improved the `README.md`.

Changes since v6:
- Rebased on top of v2.34.0.
- Inserted a commit that adds `contrib/scalar/README.md`, containing the roadmap of what I have planned for Scalar.
- The Scalar test's definition of `GIT_TEST_MAINT_SCHEDULER` has been adjusted to accommodate for a change in v2.32.0..v2.34.0.
- The config setting defaults now include `fetch.showForcedUpdates=false`, which has been identified as helping with a performance issue in large repositories.
- To avoid mistaking the current patch series for being feature-complete enough to unleash onto end users, I moved the Makefile rules to build HTML/manual pages to a later patch series.
- The patch that adds support for `-c <key>=<value>` and `-C <directory>` was moved to its own add-on patch series: While it is obvious that those options are valuable to have, an open question is whether there are other "pre-command" options in `git` that would be useful, too, and I would like to postpone _that_ discussion to that date.
- I added two patches that I had planned on keeping in an add-on patch series for later, to build and test Scalar as part of the CI. I am still not 100% certain that it is a good idea to do so already now, but let's see what the reviewers have to say.

Changes since v5:
- Fixed the commit message talking about `make -C contrib/scalar/Makefile`.
- Fixed the `git ls-tree` invocation suggested in the manual for `scalar clone`.
- Invoking `make -C contrib/scalar`, then changing a source file of `libgit.a` and then _immediately_ invoking `make -C contrib/scalar` again will now implicitly rebuild `libgit.a`.

Changes since v4:
- `scalar delete` now refuses to delete anything if it was started from within the enlistment.
- `scalar delete` releases any handles to the object store before deleting the enlistment.
- The `OBJECTS` list in the `Makefile` will now include Scalar.
- `scalar register` now supports secondary worktrees, in addition to the primary worktree.

Changes since v3:
- Moved the "Changes since" section to the top, to make it easier to see what changed.
- Reworded the commit message of the first patch.
- Removed the `[RFC]` prefix because I did not hear any objections against putting this into `contrib/`.

Changes since v2:
- Adjusted the description of the `list` command in the manual page , as suggested by Bagas.
- Addressed two style nits in `cmd_run()`.
- The documentation of `git reconfigure -a` was improved.

Changes since v1:
- A couple typos were fixed
- The code parsing the output of `ls-remote` was made more readable
- The indentation used in `scalar.txt` now consistently uses tabs
- We no longer hard-code `core.bare = false` when registering with Scalar

# Background

Microsoft invested a lot of effort into scaling Git to the needs of the Windows operating system source code. Based on the experience of the first approach, VFS for Git, the Scalar project was started. Scalar specifically has as its core goal to funnel all improvements into core Git.

# The present

The Scalar project provides a completely functional non-virtual experience for monorepos. But why stop there. The Scalar project was designed to be a self-destructing vehicle to allow those key concepts to be moved into core Git itself for the benefit of all. For example, partial clone, sparse-checkout, and scheduled background maintenance have already been upstreamed and removed from Scalar proper. This patch series provides a C-based implementation of the final remaining portions of the Scalar command. This will make it easier for users to experiment with the Scalar command. It will also make it substantially easier to experiment with moving functionality from Scalar into core Git, while maintaining backwards-compatibility for existing Scalar users.

The C-based Scalar has been shipped to Scalar users, and can be tested by any interested reader: https://github.com/microsoft/git/releases/ (it offers a Git for Windows installer, a macOS package and an Ubuntu package, Scalar has been included since v2.33.0.vfs.0.0).

# Next steps

Since there are existing Scalar users, I want to ensure backwards-compatibility with its existing command-line interface. Keeping that in mind, everything in this series is up for discussion.

I obviously believe that Scalar brings a huge benefit, and think that it would be ideal for all of Scalar's learnings to end up in `git clone`/`git init`/`git maintenance` eventually. It is also conceivable, however, that the `scalar` command could graduate to be a core part of Git at some stage in the future (such a decision would probably depend highly on users' feedback). See also [the discussion about the architecture of Scalar](https://lore.kernel.org/git/b67bbef4-e4c3-b6a7-1c7f-7d405902ef8b@gmail.com/), kicked off by Stolee.

On top of this patch series, I have lined up a few more:
1. Implement a `scalar diagnose` command.
2. Use the built-in FSMonitor (that patch series obviously needs to wait for FSMonitor to be integrated).
3. Modify the config machinery to be more generous about concurrent writes, say, to the user-wide config.
4. A few patches to optionally build and install `scalar` as part of a regular Git install (also teaching `git help scalar` to find the Scalar documentation

These are included in my [`vfs-with-scalar` branch thicket](https://github.com/dscho/git/commits/vfs-with-scalar). On top of that, this branch thicket also includes patches I do not plan on upstreaming, mainly because they are too specific either to VFS for Git, or they support Azure Repos (which does not offer partial clones but speaks the GVFS protocol, which can be used to emulate partial clones).

One other thing is very interesting about that `vfs-with-scalar` branch thicket: it contains a GitHub workflow which will run Scalar's quite extensive Functional Tests suite. This test suite is quite comprehensive and caught us a lot of bugs in the past, not only in the Scalar code, but also core Git.

# Epilogue

Now, to address some questions that I imagine every reader has who made it this far:

* Why not put the Scalar functionality directly into core Git, even a built-in? I wanted to provide an easy way for Git contributors to "play with" Scalar, without forcing a new top-level command into Git.
* Why implement the Scalar command in the Git code base? Apart from simplifying Scalar maintenance in the Microsoft port of Git, the tight version coupling between Git and Scalar reduces the maintenance burden even further. Besides, I believe that it will make it much easier to shift functionality from Scalar into core Git, once we took the hurdle of accepting the Scalar code into the code base.
* Why contribute Scalar to the Git project? We are biased, of course, yet our data-driven approach provides evidence that Scalar helps handling huge repositories with ease. By contributing it to the core Git project, we are able to share it with more users, especially some users who do not want to install Microsoft's fork of Git. We also hope that a lot of Scalar (maybe all of it) will end up in core Git, to benefit even more users.

cc: Derrick Stolee <stolee@gmail.com>
cc: Eric Sunshine <sunshine@sunshineco.com>
cc: Ævar Arnfjörð Bjarmason <avarab@gmail.com>
cc: Elijah Newren <newren@gmail.com>
cc: Bagas Sanjaya <bagasdotme@gmail.com>
cc: "Theodore Ts'o" <tytso@mit.edu>
cc: Matt Rogers <mattr94@gmail.com>
cc: Jeff King <peff@peff.net>

### CODE REVIEW

File: contrib/scalar/t/t9099-scalar.sh [hunk 1]
```diff
@@ -10,6 +10,9 @@ PATH=$(pwd)/..:$PATH
 
 . ../../../t/test-lib.sh
 
+GIT_TEST_MAINT_SCHEDULER="crontab:test-tool crontab ../cron.txt"
```
Comment (derrickstolee) [inline contrib/scalar/t/t9099-scalar.sh:13]: In the commit message, maybe include a link to https://github.com/microsoft/git/blob/vfs-2.32.0/.github/workflows/scalar-functional-tests.yml and describe that the .NET functional tests validate creating clones against a real server and then operating on them as an end-to-end scenario test.

Comment (derrickstolee) [reply contrib/scalar/t/t9099-scalar.sh:13]: Those tests also assume the functionality that exists in microsoft/git, so it is not appropriate to include such a workflow directly into git.git.

File: contrib/scalar/scalar.c [hunk 2]
```diff
@@ -392,7 +395,9 @@ static int cmd_clone(int argc, const char **argv)
 
 	if (set_config("remote.origin.url=%s", url) ||
 	    set_config("remote.origin.fetch="
-		       "+refs/heads/*:refs/remotes/origin/*") ||
+		    "+refs/heads/%s:refs/remotes/origin/%s",
+		    single_branch ? branch : "*",
+		    single_branch ? branch : "*") ||
```
Comment (derrickstolee) [inline contrib/scalar/scalar.c]: nit: leading whitespace seems incorrect.

File: contrib/scalar/scalar.c [hunk 3]
```diff
@@ -473,6 +473,69 @@ static int cmd_register(int argc, const char **argv)
...
+	strbuf_addstr(&buf, N_("scalar run <task> [<enlistment>]\nTasks:\n"));
+	for (i = 0; tasks[i].arg; i++)
+		strbuf_addf(&buf, "\t%s\n", tasks[i].arg);
+	usagestr[0] = buf.buf;
+
+	argc = parse_options(argc, argv, NULL, options,
+			     usagestr, 0);
+
+	if (argc == 0)
+		usage_with_options(usagestr, options);
```
Comment (derrickstolee) [inline contrib/scalar/scalar.c:620]: The `options` and `usagestr` are both bogus. Perhaps drop this entirely, since we don't use that functionality?

File: contrib/scalar/scalar.c [hunk 4]
```diff
@@ -684,6 +684,9 @@ int cmd_main(int argc, const char **argv)
 		argv++;
 		argc--;
 
+		if (!strcmp(argv[0], "config"))
+			argv[0] = "reconfigure";
+
```
Comment (derrickstolee) [inline contrib/scalar/scalar.c:798]: Is this really what the old `scalar config` would do? I thought there was a Scalar-specific config file somewhere (based on the `gvfs config` command, which did things like set the custom upgrader details). Maybe just drop this one?

File: contrib/scalar/scalar.c [hunk 5]
```diff
@@ -315,6 +315,45 @@ static char *remote_default_branch(const char *url)
 	return NULL;
 }
 
+static void strbuf_parentdir(struct strbuf *buf)
+{
+	int len = buf->len;
+	while (len > 0 && !is_dir_sep(buf->buf[--len]))
+		; /* keep looking for parent directory */
+	strbuf_setlen(buf, len);
```
Comment (derrickstolee) [inline contrib/scalar/scalar.c:323]: This loop appears earlier in the series, perhaps unify on the single implementation?

File: contrib/scalar/scalar.txt [hunk 6]
```diff
@@ -129,6 +130,14 @@ With the `--all` option, all enlistments currently registered with Scalar
 will be reconfigured. This option is meant to to be run every time Scalar
 was upgraded.
 
+Delete
+~~~~~~
+
+delete <enlistment>::
+    This command lets you delete an existing Scalar enlistment from your
+    local file system, unregistering the repository and stopping any file
+    watcher daemons (FSMonitor).
```
Comment (derrickstolee) [inline contrib/scalar/scalar.txt:149]: Remove this bit about FS Monitor currently and discuss it as a potential future option in the commit message.

File: contrib/scalar/scalar.txt [hunk 7]
```diff
@@ -35,6 +35,16 @@ The command implements various subcommands, and different options depending
 on the subcommand. With the exception of `clone` and `list`, all subcommands
 expect to be run in an enlistment.
 
+The following options can be specified _before_ the subcommand:
```
Comment (derrickstolee) [inline contrib/scalar/scalar.txt:38]: We should be consistent of the use of `command`/`subcommand`/`verb` throughout the series. I think you've used `command` before, which seems most correct.

File: contrib/scalar/scalar.c [hunk 8]
```diff
@@ -770,6 +770,25 @@ int cmd_main(int argc, const char **argv)
 	struct strbuf scalar_usage = STRBUF_INIT;
 	int i;
 
+	while (argc > 1 && *argv[1] == '-') {
```
Comment (derrickstolee) [inline contrib/scalar/scalar.c:773]: nit: the commit message uses back-ticks `git` when the typical Git style is to use single ticks ('git').

**Review by derrickstolee — COMMENTED** [review]
Really well organized! I found a few points that would be good to fix before sending upstream.

File: contrib/scalar/scalar.c [hunk 9]
```diff
@@ -191,9 +191,34 @@ static int toggle_maintenance(int enable)
 	return run_git("maintenance", enable ? "start" : "unregister", NULL);
 }
 
+static int add_or_remove_enlistment(int add)
```
Comment (jeffhostetler) [inline contrib/scalar/scalar.c:199]: Does add_or_remove_enlistment() belong in this commit?  I wasn't expecting it here.

Comment (jeffhostetler) [reply contrib/scalar/scalar.c:199]: Likewise for the changes in register_dir() and unregister_dir().

File: contrib/scalar/scalar.c [hunk 10]
```diff
@@ -239,6 +239,197 @@ static int unregister_dir(void)
...
+		/* Strip suffix `.git`, if any */
+		strbuf_strip_suffix(&buf, ".git");
+
+		enlistment = find_last_dir_sep(buf.buf);
+		if (!enlistment) {
+			die(_("cannot deduce worktree name from '%s'"), url);
+		}
+		enlistment = xstrdup(enlistment + 1);
+	} else {
+		usage_msg_opt(N_("need a URL"), clone_usage, clone_options);
```
Comment (jeffhostetler) [inline contrib/scalar/scalar.c:404]: I'm assuming that "URL" here refers to the "&lt;repo&gt;" from the usage.  I'm wondering about a better wording here.

File: contrib/scalar/scalar.c [hunk 11]
```diff
@@ -239,6 +239,197 @@ static int unregister_dir(void)
...
+}
+
+static int cmd_clone(int argc, const char **argv)
+{
+	const char *branch = NULL;
+	int no_fetch_commits_and_trees = 0, full_clone = 0;
+	struct option clone_options[] = {
+		OPT_STRING('b', "branch", &branch, N_("<branch>"),
+			   N_("branch to checkout after clone")),
+		OPT_BOOL(0, "no-fetch-commits-and-trees",
```
Comment (jeffhostetler) [inline contrib/scalar/scalar.c:364]: Don't OPT_BOOL options want to be the positive form and implicitly inherit the "--no-" version?
This won't set us up for "--no-no-" will it??

File: contrib/scalar/scalar.c [hunk 12]
```diff
@@ -239,6 +239,197 @@ static int unregister_dir(void)
...
+
+	if (chdir(dir) < 0) {
+		res = error_errno(_("could not switch to '%s'"), dir);
+		goto cleanup;
+	}
+
+	setup_git_directory();
+
+	/* common-main already logs `argv` */
+	trace2_data_string("scalar", the_repository, "dir", dir);
```
Comment (jeffhostetler) [inline contrib/scalar/scalar.c:432]: There is a trace2_def_repo() if that helps.   See `repo_set_worktree()`.  I cannot tell from this view if that has already been called or not.

File: Makefile [hunk 13]
```diff
@@ -2447,6 +2447,10 @@ endif
 .PHONY: objects
 objects: $(OBJECTS)
 
+SCALAR_SOURCES := contrib/scalar/scalar.c
```
[the remark this answers is not in the document]
Comment (dscho) [reply Makefile:2450]: I would like to keep the VFS for Git part in the commit message for now, if only to give readers on the Git mailing list a chance to learn about Scalar's history (so that they gain context that this has been a multi-year effort and the design is actually quite mature).

I am totally willing to cut out that part in response to your review on the Git mailing list ;-)

File: contrib/scalar/scalar.c [hunk 14]
```diff
@@ -0,0 +1,36 @@
+/*
+ * This is a port of Scalar to C.
+ */
```
[the remark this answers is not in the document]
Comment (dscho) [reply contrib/scalar/scalar.c:3]: Good idea!

File: Makefile [hunk 15]
```diff
@@ -2586,6 +2590,10 @@ $(REMOTE_CURL_PRIMARY): remote-curl.o http.o http-walker.o GIT-LDFLAGS $(GITLIBS
 	$(QUIET_LINK)$(CC) $(ALL_CFLAGS) -o $@ $(ALL_LDFLAGS) $(filter %.o,$^) \
 		$(CURL_LIBCURL) $(EXPAT_LIBEXPAT) $(LIBS)
 
+contrib/scalar/scalar$X: $(SCALAR_OBJECTS) GIT-LDFLAGS $(GITLIBS)
+	$(QUIET_LINK)$(CC) $(ALL_CFLAGS) -o $@ $(ALL_LDFLAGS) \
+		$(filter %.o,$^) $(LIBS)
+
```
[the remark this answers is not in the document]
Comment (dscho) [reply Makefile:2596]: I am not so sure. The `INCLUDE_SCALAR` knob is not even part of this patch series, and even after introducing it, users can manually run `make contrib/scalar/scalar` and it will work _without_ specifying `INCLUDE_SCALAR=OhYes`.

Would you agree that it makes sense to restrict that knob to the meta rules like `all`, `docs`, etc?

File: contrib/scalar/scalar.txt [hunk 16]
```diff
@@ -0,0 +1,36 @@
...
+Scalar is an opinionated repository management tool. By creating new
+repositories or registering existing repositories with Scalar, your Git
+experience will speed up. Scalar sets advanced Git config settings,
+maintains your repositories in the background, and helps reduce data sent
+across the network.
+
+An important Scalar concept is the enlistment: this is the top-level directory
+of the project. It contains the subdirectory `src/` which is a Git worktree.
+This encourages the separation between tracked files (inside `src/`) and
+untracked files (outside `src/`).
```
[the remark this answers is not in the document]
Comment (dscho) [reply contrib/scalar/scalar.txt:32]: I like it!

File: contrib/scalar/Makefile [hunk 17]
```diff
@@ -36,7 +36,7 @@ include ../../config.mak.uname
 TARGETS = scalar$(X) scalar.o
 GITLIBS = ../../common-main.o ../../libgit.a ../../xdiff/lib.a
 
-all: scalar$X
+all: scalar$X ../../bin-wrappers/scalar
```
[the remark this answers is not in the document]
Comment (dscho) [reply contrib/scalar/Makefile:39]: I only changed "CLI" to "command", but took the remainder verbatim.

File: contrib/scalar/scalar.c [hunk 18]
```diff
@@ -5,11 +5,197 @@
 #include "cache.h"
 #include "gettext.h"
 #include "parse-options.h"
+#include "config.h"
```
[the remark this answers is not in the document]
Comment (dscho) [reply contrib/scalar/scalar.c:8]: I changed all the commit messages in this patch series to use that format. Thank you for the suggestion.

File: contrib/scalar/scalar.c [hunk 19]
```diff
@@ -5,11 +5,197 @@
...
+
+static void setup_enlistment_directory(int argc, const char **argv,
+				       const char * const *usagestr,
+				       const struct option *options)
+{
+	if (startup_info->have_repository)
+		BUG("gitdir already set up?!?");
```
[the remark this answers is not in the document]
Comment (dscho) [reply contrib/scalar/scalar.c:17]: > Is this necessary? Perhaps replace with `assert(!startup_info->have_repository);` so debug builds will catch it?

I vastly prefer explicit `BUG()` calls over `assert()` calls, as I am totally unaware of anyone running with debug builds where the latter would be anything but a noop.

> Has this already taken @vdye's contributions in [microsoft#402](https://github.com/microsoft/git/pull/402)? Maybe it should, at least in the full non-RFC version.

No, those contributions have not yet been considered. TBH I am unsure what to do about `scalar register` in Git worktrees whose path does not end in `/src`. I was under the impression that that is not even supported, and it sure muddles an otherwise clean concept of what a Scalar enlistment is...

File: contrib/scalar/scalar.c [hunk 20]
```diff
@@ -5,11 +5,197 @@
 #include "cache.h"
 #include "gettext.h"
 #include "parse-options.h"
+#include "config.h"
+
+static void setup_enlistment_directory(int argc, const char **argv,
```
[the remark this answers is not in the document]
Comment (dscho) [reply contrib/scalar/scalar.c:12]: Right. But then, I _thought_ I knew what those rules were when I was surprised to learn that the Git worktree's name should be allowed to differ from `src`...

Could you clarify for me what a Scalar enlistment could be, other than the parent directory of a Git worktree called `src`?

File: contrib/scalar/scalar.c [hunk 21]
```diff
@@ -5,11 +5,197 @@
...
+		{ "maintenance.prefetch.schedule", "hourly" },
+		{ "maintenance.commit-graph.enabled", "true" },
+		{ "maintenance.commit-graph.auto", "0" },
+		{ "maintenance.commit-graph.schedule", "hourly" },
+		{ "maintenance.loose-objects.enabled", "true" },
+		{ "maintenance.loose-objects.auto", "0" },
+		{ "maintenance.loose-objects.schedule", "daily" },
+		{ "maintenance.incremental-repack.enabled", "true" },
+		{ "maintenance.incremental-repack.auto", "0" },
+		{ "maintenance.incremental-repack.schedule", "daily" },
```
[the remark this answers is not in the document]
Comment (dscho) [reply contrib/scalar/scalar.c:150]: I guess it would be clearer if we let `git maintenance` do the honors.

File: contrib/scalar/scalar.c [hunk 22]
```diff
@@ -5,11 +5,197 @@
...
+	/*
+	 * The `log.excludeDecoration` setting is special because we want to
+	 * set multiple values.
+	 */
+	if (git_config_get_string("log.excludeDecoration", &value)) {
+		trace2_data_string("scalar", the_repository,
+				   "log.excludeDecoration", "created");
+		if (git_config_set_multivar_gently("log.excludeDecoration",
+						   "refs/scalar/*",
+						   CONFIG_REGEX_NONE, 0) ||
```
[the remark this answers is not in the document]
Comment (dscho) [reply contrib/scalar/scalar.c:179]: Right, you mentioned that earlier and it is still in my ever-growing TODO list. Now is the exact right moment to strike that entry from that list.

File: contrib/scalar/scalar.c [hunk 23]
```diff
@@ -166,13 +186,26 @@ static int set_recommended_config(void)
 	return 0;
 }
 
+static int toggle_maintenance(int enable)
+{
+	return run_git("maintenance", enable ? "start" : "unregister", NULL);
+}
```
[the remark this answers is not in the document]
Comment (dscho) [reply contrib/scalar/scalar.c:197]: Makes sense.

File: contrib/scalar/scalar.txt [hunk 24]
```diff
@@ -38,6 +39,14 @@ register [<enlistment>]::
     `<enlistment>` is not provided, then the enlistment associated with the
     current working directory is registered.
 
+Unregister
+~~~~~~~~~~
+
+unregister [<enlistment>]::
+    Remove the specified repository from the list of repositories
+    registered with Scalar. This stops the scheduled maintenance and the
+    built-in FSMonitor.
```
[the remark this answers is not in the document]
Comment (dscho) [reply contrib/scalar/scalar.txt:112]: Yes!

File: contrib/scalar/scalar.txt [hunk 25]
```diff
@@ -26,11 +27,20 @@ This encourages the separation between tracked files (inside `src/`) and
...
 COMMANDS
 --------
 
+List
+~~~~
+
+list::
+    To see which repositories are currently registered by the service, run
+    `scalar list`. This command, like `clone`, does not need to be run
+    inside a Git worktree.
```
[the remark this answers is not in the document]
Comment (dscho) [reply contrib/scalar/scalar.txt:96]: Good point. A victim of my reordering.

File: contrib/scalar/scalar.txt [hunk 26]
```diff
@@ -27,12 +29,41 @@ This encourages the separation between tracked files (inside `src/`) and
...
+subdirectories outside your sparse-checkout by using `git ls-tree HEAD`.
+
+-b <name>::
+--branch <name>::
+	Instead of checking out the branch pointed to by the cloned repository's
+    HEAD, check out the `<name>` branch instead.
+
+--[no-]full-clone::
+    A sparse-checkout is initialized by default. This behavior can be turned
+    off via `--no-full-clone`.
```
[the remark this answers is not in the document]
Comment (dscho) [reply contrib/scalar/scalar.txt:83]: D'oh!

File: contrib/scalar/scalar.txt [hunk 27]
```diff
@@ -27,12 +29,41 @@ This encourages the separation between tracked files (inside `src/`) and
...
+
+--[no-]fetch-commits-and-trees::
+    The default behavior is to clone all commit and tree objects, with blob
+    objects being fetched on demand. With the `--no-fetch-commits-and-trees`
+    option, commit and tree objects are also fetched only as needed.
```
[the remark this answers is not in the document]
Comment (dscho) [reply contrib/scalar/scalar.txt:88]: > This option might be restricted to the GVFS protocol which has a way of downloading just the tip commit and its trees.

Woooow. I missed that. Good eyes!

File: contrib/scalar/t/t9099-scalar.sh (hunk 1 above)
Comment (dscho) [reply contrib/scalar/t/t9099-scalar.sh:13]: 👍

File: contrib/scalar/scalar.c (hunk 2 above)
Comment (dscho) [reply contrib/scalar/scalar.c]: Well spotted!

File: contrib/scalar/scalar.c (hunk 3 above)
Comment (dscho) [reply contrib/scalar/scalar.c:620]: We need them, though, to call `setup_enlistment_directory()`. So I'd rather keep this, to allow for easy addition of options in the future.

File: contrib/scalar/scalar.c (hunk 4 above)
Comment (dscho) [reply contrib/scalar/scalar.c:798]: Okay, I'll drop it.

File: contrib/scalar/scalar.c (hunk 5 above)
Comment (dscho) [reply contrib/scalar/scalar.c:323]: Makes sense. We will do that as part of incorporating https://github.com/microsoft/git/pull/402, I think.

Comment (dscho) [reply contrib/scalar/scalar.c:323]: Actually, I realized that you were talking about the code in `set_recommended_config()`, and did as  you suggested.

File: contrib/scalar/scalar.txt (hunk 6 above)
Comment (dscho) [reply contrib/scalar/scalar.txt:149]: Done.

File: contrib/scalar/scalar.txt (hunk 7 above)
Comment (dscho) [reply contrib/scalar/scalar.txt:38]: Actually, I start by detailing that "The command" (by which I mean `scalar`) implements "various subcommands". So I think I'll settle on "subcommand".

File: contrib/scalar/scalar.c (hunk 8 above)
Comment (dscho) [reply contrib/scalar/scalar.c:773]: Hmm. It is _my_ common commit message style to use backticks where code, or code-like text, is involved.

File: contrib/scalar/scalar.c (hunk 9 above)
Comment (dscho) [reply contrib/scalar/scalar.c:199]: You're right. I moved the respective code into the `register` and `unregister` commits.

File: contrib/scalar/scalar.c (hunk 10 above)
Comment (dscho) [reply contrib/scalar/scalar.c:404]: I'll borrow `_("You must specify a repository to clone.")` from `builtin/clone.c`.

File: contrib/scalar/scalar.c (hunk 11 above)
Comment (dscho) [reply contrib/scalar/scalar.c:364]: No, `OPT_BOOL()` handles `no-` specially and strips it instead of using `no-no-` for the negative form.

But that's moot now because the option does not make sense in the non-GVFS part of the patches, as pointed out by @derrickstolee (and therefore I removed it).

File: contrib/scalar/scalar.c (hunk 12 above)
Comment (dscho) [reply contrib/scalar/scalar.c:432]: Thank you for pointing me to that function, I had not noticed it yet.

File: Makefile (hunk 15 above)
[the remark this answers is not in the document]
Comment (derrickstolee) [reply Makefile:2596]: I was looking at this commit-by-commit and was expecting the `INCLUDE_SCALAR` to appear later. If you don't think that is valuable to include in this series, then ignore my comment.

File: contrib/scalar/scalar.c (hunk 8 above)
Comment (derrickstolee) [reply contrib/scalar/scalar.c:763]: Ok. Use your style.

File: contrib/scalar/scalar.c (hunk 19 above)
[the remark this answers is not in the document]
Comment (derrickstolee) [reply contrib/scalar/scalar.c:27]: The idea of `scalar register` is that we can apply our recommended config and start background maintenance on an existing Git repository that was created by a normal `git clone`. We can't adjust the partial clone or sparse-checkout for such a repo, but `scalar register` still have _some_ value there.

File: contrib/scalar/scalar.c (hunk 20 above)
[the remark this answers is not in the document]
Comment (derrickstolee) [reply contrib/scalar/scalar.c:22]: The idea of "a Scalar enlistment" is not incredibly important anymore. It _was_ in VFS for Git land where there is a `.gvfs` directory adjacent to `src`. We really think of "Git repositories registered with Scalar" as a unit, and the `src` thing is really limited to "Git repositories created with `scalar clone`". If we see the `src` pattern, then we want to assist users in deleting the repository as a complete unit.

[the remark this answers is not in the document]
Comment (dscho) [reply contrib/scalar/scalar.c:22]: > The idea of "a Scalar enlistment" is not incredibly important anymore. It _was_ in VFS for Git land where there is a `.gvfs` directory adjacent to `src`. We really think of "Git repositories registered with Scalar" as a unit, and the `src` thing is really limited to "Git repositories created with `scalar clone`". If we see the `src` pattern, then we want to assist users in deleting the repository as a complete unit.

There are exceptions to this rule, though: the output of `scalar diagnose` is put into the enlistment root, for example. Which makes it pretty important to know _where_ that root is.

**Comment by dscho** [comment]
> 
> 
> I'm happy with your latest version. Thanks!

Sorry, I had to go through all of the commits, and decided to make extensive (although admittedly only cosmetic) changes. The most important difference is that I integrated @vdye's fix for the Scalar enlistment discovery.

I also decided to squash a couple of commits that could be perceived as "oops, let's correct this" fixups to earlier commits in the topic branch.

In other words, the shape is sufficiently different enough to merit a re-review...
